include/boost/math/special_functions/interval_sieve.hpp

```diff
@@ -0,0 +1,309 @@
+// Copyright 2020 Matt Borland and Jonathan Sorenson
+//
+// Use, modification and distribution are subject to the
+// Boost Software License, Version 1.0.
+// (See accompanying file LICENSE_1_0.txt
+// or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_MATH_SPECIAL_FUNCTIONS_INTERVAL_SIEVE_HPP
+#define BOOST_MATH_SPECIAL_FUNCTIONS_INTERVAL_SIEVE_HPP
+
+#include <boost/dynamic_bitset.hpp>
+#include <boost/multiprecision/cpp_int.hpp>
+#include <boost/math/special_functions/prime_wheel.hpp>
+#include <cmath>
+#include <memory>
+#include <array>
+#include <cstdint>
+#include <bitset>
+
+namespace boost::math::detail
+{
+
+#if defined(__MPIR_VERSION) || defined(__GNU_MP_VERSION)
+// GNU GMP C or MPIR
+inline double get_double(const mpz_t &x) noexcept
+{
+    return mpz_get_d(x);
+}
+#endif
+
+#if defined(__GNU_MP_VERSION)
+#if __has_include(<gmpxx.h>)
+// GNU GMP C++ bindings
+inline double get_double(const mpz_class &x) noexcept
+{
+    return x.get_d()
```

Comment:
hmm hmm dunno about this. Shouldn't there be a `static_cast` that does this without having to `#include` all this stuff?

Reply:
GMP and MPIR do not support casting like multiprecision does. Do you think they are worth supporting? I have not fully tested their support throughout the rest of the code base. The reason they are there is this class was originally written entirely in C-style GMP by [Dr. Sorenson ](https://www.butler.edu/directory/user/jsorenso) who was kind enough to provide it to me along with other insights.

Reply:
hmm; I think @jzmaddock will have a solution for this problem.

Reply:
Do we have tests for them?

It doesn't look to me like the template will instantiate on mpz_t anyway?

mpz_class is another matter, though it would be better not to have to #include stuff.

How about (untested!):

```
class mpz_class;  // forward declare, no #include required.

template <class I>
typename std::enable_if<std::is_same<I, mpz_class>::value, double>::type get_double(const I & x)
{
   return x.get_d();
}
```
?

Reply:
I personally wouldn't support them directly, but would support them *through* multiprecision backends.

Reply:
I think I will go with @NAThompson suggestion. I tried tests with GMP directly, and it does not come close to compiling. cpp_int is generally supported and tested with a few small bugs left. I have not tested gmp_int. I will have to read the docs for that one because I am not well versed in it.

Reply:
@mborland : It took me a while to figure all the backends out, but I think you'll get the best performance numbers if you make sure the gmp backend works.

Reply:
The gmp backend passes the tests now. cpp_int comes out a bit faster here.
```
Primes less than 2^30
int64_t: 14265 ms
cpp_int: 112723 ms
mpz_int: 122094 ms
```